### PR TITLE
Restore pandapower importer and add tests

### PR DESCRIPTION
**Pull request by pz-max**
## Changes proposed in this Pull Request
The pandapower network importer was not working correctly. @rbolgaryn and myself working currently on fixing the importer. In particular, we are fixing the transformer and the shunt components.

## Checklist

- [x] Code changes are sufficiently documented; i.e. new functions contain docstrings and further explanations may be given in `doc`.
- [x] Unit tests for new features were added (if applicable).
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included.
- [x] I consent to the release of this PR's code under the MIT license.

### CODE REVIEW

**Comment by pz-max** [comment]
Update. The AC power flow seems to match with pandapower and other tools https://github.com/G-PST/power-flow-exercise, confirming the integration. More tests will be performed before this PR is ready for review

File: doc/release_notes.rst [hunk 1]
```diff
@@ -7,7 +7,10 @@ Upcoming Release
 
 .. warning:: The features listed below are not released yet, but will be part of the next release! To use the features already you have to install the ``master`` branch, e.g. ``pip install git+https://github.com/pypsa/pypsa#egg=pypsa``.
 
-* new features
+**New Features**
+
+* PyPSA can import pandapower networks. Now, some issues regarding the transformer component and indexing are fixed. Shunts can also be imported. [`#332 <https://github.com/PyPSA/PyPSA/pull/332>`_]
```
Comment (fneum) [inline doc/release_notes.rst:12]: PyPSA could import from pandapower for earlier versions. Maybe rephrase to: "restore import of pandapower networks".

File: pypsa/io.py [hunk 2]
```diff
@@ -1010,7 +1009,14 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
...
     """
     logger.warning("Warning: Importing from pandapower is still in beta; not all pandapower data is supported.\nUnsupported features include: three-winding transformers, switches, in_service status, shunt impedances and tap positions of transformers.")
-
+    
+    if use_pandapower_index:
+        for element in ["bus", "line", "gen", "load", "shunt", "trafo", "switch"]:
+            net[element].name = net[element].index.values
+        
+        net.sgen.name = net.sgen.index.values + (net.gen.index.max() + 1 if len(net.gen) else 0)
+        net.ext_grid.name = net.ext_grid.index.values + (net.gen.index.max() + 1 if len(net.gen) else 0) + (net.sgen.index.max() + 1 if len(net.sgen) else 0)
```
Comment (fneum) [inline pypsa/io.py:1018]: not sure I understand what this segment does.

**Review by fneum — CHANGES_REQUESTED** [review]
Great, that'll be useful. Could you add a unit test for a LOPF that checks pandapower and PyPSA with imported pandapower network produce the same results?

File: pypsa/io.py [hunk 3]
```diff
@@ -1019,38 +1025,38 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
 
     d["Load"] = pd.DataFrame({"p_set" : (net.load.scaling*net.load.p_mw).values,
                               "q_set" : (net.load.scaling*net.load.q_mvar).values,
-                              "bus" : net.bus.name.loc[net.load.bus].values},
+                              "bus" : net.bus.name.loc[net.load.bus.values].values},
```
[the remark this answers is not in the document]
Comment (pz-max) [reply pypsa/io.py:1028]: Not sure. However, works :+1: 
Could test if we can drop it

[the remark this answers is not in the document]
Comment (fneum) [reply pypsa/io.py:1028]: Yes, please test if it works without. Would be nicer.

File: pypsa/io.py [hunk 4]
```diff
@@ -1019,38 +1025,38 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
...
+    d["Bus"].loc[net.bus.name.loc[net.gen.bus.values].values,"v_mag_pu_set"] = net.gen.vm_pu.values
 
 
     #deal with PQ "static" generators
-    d["Generator"] = pd.concat((d["Generator"],pd.DataFrame({"p_set" : -(net.sgen.scaling*net.sgen.p_mw).values,
-                                                             "q_set" : -(net.sgen.scaling*net.sgen.q_mvar).values,
-                                                             "bus" : net.bus.name.loc[net.sgen.bus].values,
+    d["Generator"] = pd.concat((d["Generator"],pd.DataFrame({"p_set" : (net.sgen.scaling*net.sgen.p_mw).values,
+                                                             "q_set" : (net.sgen.scaling*net.sgen.q_mvar).values,
+                                                             "bus" : net.bus.name.loc[net.sgen.bus.values].values,
```
[the remark this answers is not in the document]
Comment (fneum) [reply pypsa/io.py:1044]: in 2.0.0

[the remark this answers is not in the document]
Comment (fneum) [reply pypsa/io.py:1044]: https://github.com/e2nIEE/pandapower/issues/208

File: pypsa/io.py (hunk 2 above)
Comment (pz-max) [reply pypsa/io.py:1018]: This segment takes care of unique naming of all assets. Pandapower splits sgen, gen and ext_grid (slack generator). To guarantee unique bus names required naming changes in pandapower. Line 1017-1018 just add a running number to the name.

**Comment by pz-max** [comment]
> Great, that'll be useful. Could you add a unit test for a LOPF that checks pandapower and PyPSA with imported pandapower network produce the same results?

Yes, I can write a unit test for the pandapower importer. :)
I would create a [minimal pandapower network](https://www.pandapower.org/start/#a-short-introduction-), import it to pypsa and test the powerflow for both models.

**Comment by fneum** [comment]
Yes, or OPF https://github.com/e2nIEE/pandapower/blob/master/tutorials/opf_basic.ipynb

File: pypsa/io.py (hunk 2 above)
Comment (FabianHofmann) [reply pypsa/io.py:1018]: This is a bit hacky. You change the dataframes of the pandapower network which should not happen (the pandapower network is only a input and should leave the function as it came in). Do you really think this is necessary?

File: pypsa/io.py [hunk 5]
```diff
@@ -1019,38 +1025,38 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
...
                                                              "control" : "PQ"},
                                                             index=net.sgen.name)), sort=False)
 
     d["Generator"] = pd.concat((d["Generator"],pd.DataFrame({"control" : "Slack",
                                                              "p_set" : 0.,
                                                              "q_set" : 0.,
-                                                             "bus" : net.bus.name.loc[net.ext_grid.bus].values},
+                                                             "bus" : net.bus.name.loc[net.ext_grid.bus.values].values},
                                                             index=net.ext_grid.name.fillna("External Grid"))), sort=False)
```
Comment (FabianHofmann) [inline pypsa/io.py:1052]:
```suggestion
                                                             "bus" : net.bus.name.loc[net.ext_grid.bus.values].values},
                                                            index=net.ext_grid.name.fillna("External Grid"))), sort=False,
                                                            ignore_index=ignore_generator_index)
```

**Comment by pz-max** [comment]
> Yes, or OPF https://github.com/e2nIEE/pandapower/blob/master/tutorials/opf_basic.ipynb

Currently, we only tested successfully the powerflow for the [RTS_GMLC network](https://github.com/G-PST/power-flow-exercise/blob/f8a37ce06667fbc81195fe4af42aeedca6993e84/example-pypsa/pypsa-example.py#L62-L78).

A trial to make the [simple_network](https://pandapower.readthedocs.io/en/v2.0.0/networks/example.html#simple-example-network) of pandapower failed. Probably related to the non-supported switch. Adding these features might be good for a good pandapower <-> pypsa compatibility. This needs some work which I can't tackle atm.

However, I will give the OPF example next week a try. Hopefully, this works. If not, I will create manually a pandapower example. Probably the best solution to avoid non-supported features.

**Comment by fneum** [comment]
No ambition to include switches atm. Didn’t know you only tested PF. Maybe create an example without switches then?

File: pypsa/io.py [hunk 6]
```diff
@@ -1088,18 +1094,19 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
     else:
         s_nom = net.trafo.sn_mva.values/1000.
```
Comment (dannerph) [inline pypsa/io.py:1095]: s_nom is given in MVA, net.trafo.sn_mva is already in MVA, no need to divide by 1000.

**Comment by pz-max** [comment]
Picking this PR up for the weekend...

**Plan:**
1. Add fixes as discussed above
2. Build unit test by building an own simple pandapower network 

**What is not tackled but might be interesting for anyone else:**
- Make sure that PyPSA can import all the example networks from pandapower (IEEE networks etc.). This might require the addition of new components in PyPSA like switches.

File: pypsa/io.py (hunk 6 above)
Comment (pz-max) [reply pypsa/io.py]: fixed now thanks

File: pypsa/io.py (hunk 5 above)
Comment (pz-max) [reply pypsa/io.py]: using now `ignore_index` when concat all df's

File: pypsa/io.py (hunk 2 above)
Comment (pz-max) [reply pypsa/io.py]: removed these lines

**Comment by pz-max** [comment]
Hi @fneum @FabianHofmann , I think this PR is ready for review. I addressed the previous discussed points.
What this PR does:
- [x] restore the pandapower import such that the pf is now matching
- [x] import now also shunts and fixes transformer formulation
- [x] add tests for pandapower network imports
- [x] add test of imported pandapower "power flow"

What the PR does not:
- [ ] test the powerflow when q_mw of shunts > 0 (there is a bug somewhere, that needs work - powerflow comparison between pandapower and pypsa is not matching)
- [ ] test OPF between pandapower and pypsa (running out of time, could be a nice exercise)


Here are some tips if someone wants to catch up on this work:

Create a fork on github, clone it, and checkout locally to another branch:
```
git clone https://github.com/pz-max/PyPSA.git
cd PyPSA
git checkout -b <name of new branch>  # custom branch name e.g. `pp-opf`
conda env create --f environment.yaml  # install all tools for pypsa
conda activate pypsa
conda env update -f environment_dev.yaml  # updates env with dev tools e.g. pytest etc.
pip install -e .  # create pypsa locally in developer mode
```

With this setup you can run `pytest` and also use a local debugger.
E.g. to debug the `pypsa/io.py` this might be helpful to add it at the bottom of the script
while setting breakpoints at some spots across the script:
```
if __name__ == "__main__":
    import pandapower as pp
    import pandapower.networks as pn
    import pypsa
    network = pypsa.Network()
    net = pp.create_empty_network()
    
    # custom network
    # bus1 = pp.create_bus(net, vn_kv=20., name="Bus 1")
    # bus2 = pp.create_bus(net, vn_kv=0.4, name="Bus 2")
    # bus3 = pp.create_bus(net, vn_kv=0.4, name="Bus 3")
    # pp.create_ext_grid(net, bus=bus1, vm_pu=1.02, name="Grid Connection")
    # pp.create_load(net, bus=bus3, p_mw=0.100, q_mvar=0.05, name="Load")
    # pp.create_shunt(net, bus=bus3, p_mw=0, q_mvar=0.005, name='Shunt')
    # pp.create_transformer(net, hv_bus=bus1, lv_bus=bus2, std_type="0.4 MVA 20/0.4 kV", name="Trafo")
    # pp.create_line(net, from_bus=bus2, to_bus=bus3, length_km=0.1, std_type="NAYY 4x50 SE", name="Line")
    # CIGRE network
    net = pn.create_cigre_network_mv(with_der='all')
    network.import_from_pandapower_net(net)
    print("hello world")
```

File: pypsa/io.py [hunk 7]
```diff
@@ -1308,15 +1305,22 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
...
+        x = np.sqrt(z**2 - r**2)
+
+        y = net.trafo.i0_percent.values / 100.0
+        g = (
+            net.trafo.pfe_kw.values
+            / net.trafo.sn_mva.values
+            / 1000
+            * 1.0
+            / net.trafo.sn_mva.values
+        )
```
Comment (FabianHofmann) [inline pypsa/io.py:1322]:
```suggestion
        y = net.trafo.i0_percent.values / 100.0
        g = (
            net.trafo.pfe_kw.values
            / net.trafo.sn_mva.values
            / 1000
            / net.trafo.sn_mva.values
        )
```

**Review by FabianHofmann — CHANGES_REQUESTED** [review]
@pz-max this looks very good! I now just reviewed the code, just a few remarks. I did not test it interactively but the tests seem to be covering a lot.

File: pypsa/io.py [hunk 8]
```diff
@@ -1308,15 +1305,22 @@ def import_from_pandapower_net(network, net, extra_line_data=False):
...
-        s_nom = net.trafo.sn_mva.values / 1000.0
-
-        r = net.trafo.vkr_percent.values / 100.0
-        x = np.sqrt((net.trafo.vk_percent.values / 100.0) ** 2 - r**2)
-        # NB: b and g are per unit of s_nom
-        g = net.trafo.pfe_kw.values / (1000.0 * s_nom)
-
-        # for some bizarre reason, some of the standard types in pandapower have i0^2 < g^2
-        b = -np.sqrt(((net.trafo.i0_percent.values / 100.0) ** 2 - g**2).clip(min=0))
+        s_nom = net.trafo.sn_mva.values
```
[the remark this answers is not in the document]
Comment (pz-max) [reply pypsa/io.py:1308]: Pandapower uses MVA & we use MVA so there was no need to /1000

[the remark this answers is not in the document]
Comment (pz-max) [reply pypsa/io.py:1308]: vkr_percent also does not need a /1000 conversion

**Review by FabianHofmann — APPROVED** [review]

**Review by fneum — APPROVED** [review]

**Comment by FabianHofmann** [comment]
As soon as @fneum gives his approval we can merge this. @pz-max could you raise an issue about the remaining points that we don't forget about it?
